### PR TITLE
Improve the wasm idlharness test.

### PR DESCRIPTION
**Pull request by Ms2ger**
CC @foolip

### CODE REVIEW

**Review by foolip — COMMENTED** [review]
Alright, I've done it, I've reviewed!

Nice new helper functions that make it more obvious in many places what's going on than with `self[this.name]`, just some naming quibbles really.

Also, with a change this large, it seems like something will break. Can you run it through Gecko's CQ to see if any tests are broken by it?

**Comment by Ms2ger** [comment]
Fixed your comments, checked in Firefox CI, and fixed the bug I introduced in `test_member_iterable()`. Thanks for making me do that :)

**Comment by foolip** [comment]
@Ms2ger, I can't review the delta between last time and now because you force pushed the branch (I think) but I'll pay special attention in `test_member_iterable()` to see if I can spot the change.

File: resources/test/tests/unit/IdlInterface/get_interface_object.html [hunk 1]
```diff
@@ -0,0 +1,22 @@
...
+test(function() {
+    window.A = {};
+    var i = interfaceFrom('interface A { };');
+    assert_equals(i.get_interface_object(), window.A);
+}, 'Interface does not have LegacyNamespace.');
+
+test(function() {
+    window.Foo = { A: {} };
+    var i = interfaceFrom('[LegacyNamespace=Foo] interface A { }; namespace Foo { };');
+    assert_equals(i.get_interface_object(), window.Foo.A);
```
Comment (foolip) [inline resources/test/tests/unit/IdlInterface/get_interface_object.html]: Nice succinct tests!

**Review by foolip — APPROVED** [review]
Nah, couldn't spot the difference in `test_member_iterable()` from memory, but I'm sure it's good.

I thought I'd try running this through Chromium's CQ as well, which I can do if I get https://chromium-review.googlesource.com/c/chromium/src/+/1262716 landed first quickly. If it takes me beyond Friday, just land this :)

**Comment by Ms2ger** [comment]
(I forgot to add the `bind(this)` calls initially.)

**Comment by foolip** [comment]
Actually, I don't need to wait for https://chromium-review.googlesource.com/c/chromium/src/+/1262716, I confused myself. I'll try it now.

**Comment by foolip** [comment]
https://chromium-review.googlesource.com/c/chromium/src/+/1264163 (if it's all green then no need to wait for me to report back)

File: resources/idlharness.js [hunk 2]
```diff
@@ -1514,10 +1550,10 @@ IdlInterface.prototype.test_self = function()
             // "If I was not declared with a [Constructor] extended attribute,
             // then throw a TypeError."
             assert_throws(new TypeError(), function() {
-                self[this.name]();
+                this.get_interface_object()();
             }.bind(this), "interface object didn't throw TypeError when called as a function");
             assert_throws(new TypeError(), function() {
-                new self[this.name]();
+                new this.get_interface_object()();
```
Comment (foolip) [inline resources/idlharness.js]: There's a bug here. `DetectedBarcode` doesn't have a constructor in shape-detection-api.idl, but it does in Blink. A test used to fail for that but no longer does:
```diff
--- /b/s/w/ioYUXUDo/layout-test-results/external/wpt/shape-detection/idlharness.any-expected.txt
+++ /b/s/w/ioYUXUDo/layout-test-results/external/wpt/shape-detection/idlharness.any-actual.txt
@@ -7,7 +7,7 @@
 PASS FaceDetector interface: existence and properties of interface prototype object's "constructor" property
 PASS FaceDetector interface: existence and properties of interface prototype object's @@unscopables property
 PASS FaceDetector interface: operation detect(ImageBitmapSource)
-FAIL DetectedFace interface: existence and properties of interface object assert_throws: interface object didn't throw TypeError when called as a constructor function "function () { [native code] }" did not throw
+PASS DetectedFace interface: existence and properties of interface object
 PASS DetectedFace interface object length
 PASS DetectedFace interface object name
 PASS DetectedFace interface: existence and properties of interface prototype object
@@ -23,7 +23,7 @@
 PASS BarcodeDetector interface: existence and properties of interface prototype object's @@unscopables property
 FAIL BarcodeDetector interface: attribute supportedFormats assert_true: The prototype object must have a property "supportedFormats" expected true got false
 PASS BarcodeDetector interface: operation detect(ImageBitmapSource)
-FAIL DetectedBarcode interface: existence and properties of interface object assert_throws: interface object didn't throw TypeError when called as a constructor function "function () { [native code] }" did not throw
+PASS DetectedBarcode interface: existence and properties of interface object
 PASS DetectedBarcode interface object length
 PASS DetectedBarcode interface object name
 PASS DetectedBarcode interface: existence and properties of interface prototype object
```

The same has happened with the `RTCIceTransport` interface.

Unfortunately neither of those are shipped yet so you can't try against a vanilla Chrome. But could you check if perhaps the change in the condition has just made this code unreachable in this case? That would explain the incorrect pass.

**Review by foolip — COMMENTED** [review]
https://chromium-review.googlesource.com/c/chromium/src/+/1264163 has results now and it looks like there's a real regression (manifesting as fail->pass) and the message for assert_interface_object_exists could do with some tweaking.

File: resources/idlharness.js (hunk 2 above)
Comment (Ms2ger) [reply resources/idlharness.js]: The issue was operator precedence; this fixed it:
```diff
-                new this.get_interface_object()();
+                new (this.get_interface_object())();
```

but I used a slightly different fix in the end.

(I notice that this test isn't particularly useful; it only catches a constructor without required arguments. Oh well.)

**Comment by foolip** [comment]
Post-merge review, the fixup commits in https://github.com/web-platform-tests/wpt/compare/master...Ms2ger:wasm-idl-review LGTM.

The changes are being imported into Chromium now and no unexpected changes AFAICT:
https://chromium-review.googlesource.com/c/chromium/src/+/1278537

Note that one failures message went from "self.SVGPoint should be the same value as self.DOMPoint" to "self.SVGPoint should be the same value as the interface object", and the former was more clear/accurate. @Ms2ger, would you care to try to fix that?

**Pull request merged**